app/server/api/ussd_api.py

```diff
@@ -31,19 +32,24 @@ def post(self):
         user_input = post_data.get('text')
         service_code = post_data.get('serviceCode')
 
+
         if phone_number:
             msisdn = proccess_phone_number(phone_number, 'KE')
-            user = User.query.filter_by(phone=msisdn).first()
+            user = User.query.execution_options(show_all=True).filter_by(phone=msisdn).first()
+            # api chains all inputs that came through with *
+            latest_input = user_input.split('*')[-1]
             # TODO(ussd): 'exit_not_registered' if no user
             if None in [user, msisdn, session_id]:
                 current_menu = UssdMenu.find_by_name('exit_invalid_request')
                 text = menu_display_text_in_lang(current_menu, user)
             else:
-                current_menu = KenyaUssdProcessor.process_request(session_id, user_input, user)
-                ussd_session = create_or_update_session(session_id, user, current_menu, user_input, service_code)
+                current_menu = KenyaUssdProcessor.process_request(session_id, latest_input, user)
+                ussd_session = create_or_update_session(session_id, user, current_menu, latest_input, service_code)
                 text = menu_display_text_in_lang(current_menu, user)
                 #TODO(ussd): change this out for i18n placeholders
                 text = KenyaUssdProcessor.replace_vars(current_menu, ussd_session, text, user)
+
+                db.session.commit()
```

Comment:
i removed all commits and just do one at the very end instead. this seems a bit scary to not have commits anywhere else. but otherwise we get that user is deleted error.... thoughts?

Reply:
this seems fine to me lol. Why u find scary?

Reply:
also, a while back I implemented the auto commit that happens at the end of a request if it's a 200 response, so technically u don't need a commit here at all, but I'm kinda feelling like i now dislike the auto-commit for being far too magic..

Reply:
I disagree. I think it makes perfect sense to auto-commit on successful responses.

Reply:
the reason i dislike it is that it's not uncommon for us to want to commit on a failed response as well to log the failure reason in the db, and this needs to be done manually in the api function  itself. This means the way we do commits becomes inconsistent between successes and failures, which feels bad compared to "commit your session before you return a response if you want keep the data, and don't if you don't".